Arkane output fixes

goldmanm opened the pull request:
<!--
Thanks for contributing a pull request! Please try to provide as much detail as possible to help the reviewer understand your work.
You can also add the appropriate labels to describe the topic of the pull request and the type of changes you're making.
-->

### Motivation or Problem


Arkane makes calculating rate coefficients easy, but suffers from not outputting all information needed in publications and also sometimes has unexpected behavior due to poor structuring.

### Description of Changes

This PR modularizes core methods, like loading statmech, kinetics, and thermo jobs; fixes the `--plot` option; uses species labels in chemkin files; adds try-except statments to prevent output from erroring; creates methods to parse T1 diagnostics, levels of theory, and geometries; allows for termolecular products to be output by pdep; outputs more quantum information necessary for publications; and makes chemkin output and merge models more robust.

### Testing
Many of the new methods have unit tests to ensure proper functioning. I have run this version for large pressure dependent networks and highp kinetics and they function properly. I have not tested the explorer functionality, though the test methods for it continue to work. 

### Reviewer Tips
Test the explorer functionality. Ensure code makes sense and is written in a usable and expandable way. 

<!--
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with master?
-->

alongd commented:
@goldmanm, thanks!
Before the review, could you take a look at the RMG-Tests?

mliu49 commented:
FYI, to re-trigger RMG-tests, restart the travis push build.

alongd commented on arkane/gaussian.py:
```diff
@@ -413,6 +413,67 @@ def loadScanEnergies(self):
 
         return Vlist, angle
 
+    def _load_scan_specs(self,letter_spec):
```
space after comma

goldmanm replied:
added

alongd commented on arkane/statmech.py:
```diff
@@ -461,14 +461,27 @@ def load(self, pdep=False, plot=False):
+                        try:
+                            pivot_atoms = scanLog.load_scan_pivot_atoms()
+                        except Exception as e:
+                            logging.warning("Unable to find pivot atoms in scan due to error: {}".format(e))
```
log specie label

goldmanm replied:
added

alongd commented on arkane/statmech.py:
```diff
@@ -461,14 +461,27 @@ def load(self, pdep=False, plot=False):
                     if symmetry is None:
                         symmetry = determine_rotor_symmetry(v_list, self.species.label, pivots)
-                    self.raw_hindered_rotor_data.append((self.species.label, rotorCount, symmetry, angle, v_list))
+                    self.raw_hindered_rotor_data.append((self.species.label, rotorCount, symmetry, angle, v_list, pivot_atoms, frozen_atoms))
```
break the line

goldmanm replied:
fixed

alongd commented on rmgpy/tools/merge_models.py:
```diff
@@ -122,9 +122,15 @@ def execute(inputModelFiles, **kwargs):
-        print 'Added {1:d} out of {2:d} ({3:.1f}%) unique reactions from model #{0:d}.'.format(i+1, Nrxn - Nrxn0, len(model.reactions), (Nrxn - Nrxn0) * 100. / len(model.reactions))
-    
+        if  len(model.species) > 0:
+            print 'Added {1:d} out of {2:d} ({3:.1f}%) unique species from model #{0:d}.'.format(i+1, Nspec - Nspec0, len(model.species), (Nspec - Nspec0) * 100. / len(model.species))
```
break the lines

goldmanm replied:
broke the 5 long print statements.

alongd commented on arkane/gaussian.py:
```diff
@@ -497,3 +497,31 @@ def loadNegativeFrequency(self):
+            while line != '':
+                if not diagnostic_section and line.startswith(' 1\\1\\'):
+                    diagnostic_section = True
+                    print(ending_diagnostics)
```
was this line left from debugging?

goldmanm replied:
yes. removed. thank you.

alongd commented on arkane/gaussianTest.py:
```diff
@@ -160,6 +165,14 @@ def testDetermineQMSoftware(self):
+        """
+        Ensures get_level_of_theory works on gaussian log files
+        """
+        level_theory = ['CBS-QB3','G3','UB3LYP']
```
space after comma

goldmanm replied:
added

alongd commented on arkane/gaussian.py:
```diff
@@ -497,3 +497,31 @@ def loadNegativeFrequency(self):
+
+    def get_level_of_theory(self):
+        """
+        Returns a string describing the level of theory used in the
```
Does this return the method without the basis set?
I think only the method + basis set (e.g., b3lyp/6-311g) is called a level of theory, otherwise let's just call it "method"

goldmanm replied:
renamed to get_quantum_method

goldmanm commented on arkane/main.py:
```diff
@@ -125,7 +125,7 @@ def parseCommandLineArguments(self):
 
         # Add options for controlling generation of plots
-        parser.add_argument('-p', '--plot', action='store_true', default=True, help='generate plots of results')
+        parser.add_argument('-p', '--plot', action='store_true', default=False, help='generate plots of results')
```
I don't really care which is the default. This commit was just trying to make sure the attributes are functioning properly and consistent with the documentation.

alongd replied:
OK. So I think we should plot by default, and have a "no_plots" flag

goldmanm replied:
Fixed in new commit.

goldmanm commented on rmgpy/pdep/configuration.pyx:
```diff
@@ -105,7 +105,14 @@ cdef class Configuration:
+
+    cpdef bint isTermolecular(self) except -2:
+        """
+        Return ``True`` if the configuration represents a bimolecular reactant
```
fixed

goldmanm commented on rmgpy/pdep/configuration.pyx:
```diff
@@ -278,7 +285,26 @@ cdef class Configuration:
+                mu = 1.0/(1.0/mass[0] + 1.0/mass[1])
+                modes.insert(0, IdealGasTranslation(mass=(mu/constants.amu,"amu")))
+                mu2 = 1.0/(1.0/mass[0] + 1.0/mass[2])
+                modes.insert(0, IdealGasTranslation(mass=(mu2/constants.amu,"amu")))
```
I tried with one mode and three modes and both led to a significant error. The reason for two modes is that it was consistent thermodynamics with the pdep algorithm.

goldmanm replied:
Conceptually this represents six transnational degrees of freedom. One between molecule 1 and 2 and one between molecule 1 and 3.

If we counted them all as one 'molecule' we would have 6 extra vibrations, so having two translations seems to make sense from this viewpoint.

goldmanm commented on rmgpy/pdep/configuration.pyx:
```diff
@@ -278,7 +285,26 @@ cdef class Configuration:
                 mu = 1.0/(1.0/mass[0] + 1.0/mass[1])
                 modes.insert(0, IdealGasTranslation(mass=(mu/constants.amu,"amu")))
-                
+            elif self.isTermolecular():
```
Great idea! just implemented this.

goldmanm commented on rmgpy/chemkin.pyx:
```diff
@@ -76,7 +76,7 @@ def readThermoEntry(entry, Tmin=0, Tint=0, Tmax=0):
     Format specification at http://www2.galcit.caltech.edu/EDL/public/formats/chemkin.html
     """
     lines = entry.splitlines()
-    species = str(lines[0][0:18].split()[0].strip())
```
I replaced with a live link describing thermo format.

goldmanm commented on rmgpy/chemkin.pyx:
```diff
@@ -1430,11 +1430,13 @@ def getSpeciesIdentifier(species):
 ################################################################################
 
-def writeThermoEntry(species, elementCounts=None, verbose=True):
+def writeThermoEntry(species, elementCounts=None, verbose=True, use_label = False):
```
Fixed these. There are a few other functions in the file with extra spaces, though it might be odd adding those to this PR.

goldmanm commented on rmgpy/chemkin.pyx:
```diff
@@ -1524,7 +1524,7 @@ def writeReactionString(reaction, javaLibrary = False, use_label = False):
         else:
             reaction_string = ' + '.join([getSpeciesIdentifier(reactant) for reactant in reaction.reactants])
-        reaction_string += ' => ' if not reaction.reversible else ' = '
+        reaction_string += '<=>' if reaction.reversible else '=>'
```
We should definitely keep them as long as it doesn't cause issues in the length of the reaction in chemkin interpreter. Added the spaces.

mliu49 replied:
I think you missed adding spaces to this line.

goldmanm commented on arkane/thermo.py:
```diff
@@ -193,7 +193,7 @@ def save(self, outputFile):
         else:
             elementCounts = {'C': 0, 'H': 0}
-        chemkin_thermo_string = writeThermoEntry(species, elementCounts=elementCounts, verbose=True)
+        chemkin_thermo_string = writeThermoEntry(species, elementCounts=elementCounts, verbose=True, use_label = True)
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -257,7 +257,10 @@ def execute(self):
+                    f.write(job.species.label)
+                    if len(job.species.label) > 16:
+                        f.write('! species label is longer than chemkin supports. Consider shortening it.')
+                        logging.warning('Species {} has a label {} char longer than chemkin supports. Consider shortening it.'.format(job.species.label, len(job.species.label) - 16))
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -257,7 +257,10 @@ def execute(self):
-                    f.write(job.species.toChemkin())
+                    f.write(job.species.label)
+                    if len(job.species.label) > 16:
+                        f.write('! species label is longer than chemkin supports. Consider shortening it.')
```
added

goldmanm commented on rmgpy/chemkin.pyx:
```diff
@@ -1681,7 +1681,8 @@ def writeKineticsEntry(reaction, speciesList, verbose = True, javaLibrary = Fals
     reaction_string = writeReactionString(reaction, javaLibrary, use_label = use_label)
-    
+    if len(reaction_string) > 52:
+        string += '! This reaction has length {0} which is {1} longer than Chemkin allows.\n! Consider shortening species names to run chemkin.'.format(len(reaction_string), len(reaction_string) - 52)
```
added

goldmanm commented on arkane/thermo.py:
```diff
@@ -70,23 +70,33 @@ def __init__(self, species, thermoClass):
+            try:
+                self.write_output(output_directory)
+            except Exception as e:
+                logging.warning("Could not write outputfile due to error: ".format(e))
```
fixed

goldmanm commented on arkane/thermo.py:
```diff
@@ -70,23 +70,33 @@ def __init__(self, species, thermoClass):
+            try:
+                self.write_output(output_directory)
+            except Exception as e:
+                logging.warning("Could not write outputfile due to error: ".format(e))
```
added

goldmanm commented on arkane/thermo.py:
```diff
@@ -70,23 +70,33 @@ def __init__(self, species, thermoClass):
+                try:
+                    self.plot(output_directory)
+                except Exception as e:
+                    logging.warning("Could not write chemkin output due to error: ".format(e))
```
corrected to 'could not generate plots due to error'.

goldmanm commented on arkane/thermo.py:
```diff
@@ -181,8 +192,14 @@ def save(self, outputFile):
+        `species_dictionary.txt` for the outut_directory specified
+        """
+        species = self.species
+        f = open(os.path.join(output_directory, 'chem.inp'), 'a')
```
yup. fixed

goldmanm commented on arkane/statmech.py:
```diff
@@ -185,14 +185,17 @@ def __init__(self, species, path):
+            try:
+                self.write_output(output_directory)
+            except Exception as e:
+                logging.warning("Could not write statmech outputfile due to error: ".format(e))
```
fixed

goldmanm commented on arkane/statmech.py:
```diff
@@ -185,14 +185,17 @@ def __init__(self, species, path):
+            try:
+                self.write_output(output_directory)
+            except Exception as e:
+                logging.warning("Could not write statmech outputfile due to error: ".format(e))
```
fixed

goldmanm commented on arkane/statmech.py:
```diff
@@ -550,12 +553,13 @@ def load(self, pdep=False):
         at `path` on disk.
         """
 
+        outputFile = os.path.join(output_directory,'output.py')
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -273,7 +273,7 @@ def execute(self):
             if isinstance(job, ThermoJob):
                 job.execute(output_directory=self.outputDirectory, plot=self.plot)
             if isinstance(job, StatMechJob):
-                job.execute(outputFile=outputFile, plot=self.plot, pdep=is_pdep(self.jobList))
```
will fix after rebase

goldmanm commented on arkane/kinetics.py:
```diff
@@ -137,22 +137,35 @@ def Tlist(self):
+            try:
+                self.write_output(output_directory)
+            except Exception as e:
+                logging.warning("Could not write kinetics outputfile due to error: {}".format(e))
```
fixed that and separated outputfile to 2 words

goldmanm replied:
and the following two other ones.

goldmanm commented on arkane/kinetics.py:
```diff
@@ -282,19 +296,24 @@ def save(self, outputFile):
         # Reaction path degeneracy is INCLUDED in the kinetics itself!
-        rxn_str = 'kinetics(label={0!r}, kinetics={1!r})'.format(reaction.label, reaction.kinetics)
-        f.write('{0}\n\n'.format(prettify(rxn_str)))
+        string = 'kinetics(label={0!r}, kinetics={1!r})'.format(reaction.label, reaction.kinetics)
```
sorry about that. I think i missed this when trying to merge with the pep8 pull request.

goldmanm commented on arkane/kinetics.py:
```diff
@@ -282,19 +296,24 @@ def save(self, outputFile):
+        `species_dictionary.txt` for the outut_directory specified
+        """
+
+        # obtain A unit conversion factor
```
typo. fixed

goldmanm commented on arkane/main.py:
```diff
@@ -304,10 +304,8 @@ def execute(self):
-                                                                     self.jobList]):
-                # if there is an explorer job the pdep job will be run in the explorer job
+                job.execute(output_directory=self.outputDirectory, plot=self.plot)
+            elif isinstance(job, PressureDependenceJob) and not any([isinstance(job, ExplorerJob) for job in self.jobList]): #if there is an explorer job the pdep job will be run in the explorer job
```
fixed

goldmanm commented on arkane/molpro.py:
```diff
@@ -408,3 +408,18 @@ def loadScanEnergies(self):
+    def get_T1_diagnostic(self):
+        """
+        Returns the T1 diagnostic from output log.
+        If multiple occur, returns the last occurance
```
fixed

goldmanm commented on arkane/log.py:
```diff
@@ -155,3 +155,9 @@ def get_optical_isomers_and_symmetry_number(self):
+
+    def get_T1_diagnostic(self):
+        """
+        This method returns the T1 diagnostic for certain quantuum jobs
```
fixed

goldmanm commented on arkane/molpro.py:
```diff
@@ -408,3 +408,18 @@ def loadScanEnergies(self):
+        If multiple occur, returns the last occurance
+        """
+        with open(self.path) as f:
+            log = f.read().splitlines()
```
i don't think anything. switched to the simpler notation.

goldmanm commented on arkane/molpro.py:
```diff
@@ -408,3 +408,18 @@ def loadScanEnergies(self):
+            if 'T1 diagnostic:  ' in line:
+                items = line.split()
+                return float(items[-1])
+        raise ValueError('Unable to find T1 diagnostic in energy file:{}'.format(self.path))
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -274,33 +274,40 @@ def execute(self):
             if isinstance(job, StatMechJob):
                 job.execute(output_directory=self.outputDirectory, plot=self.plot, pdep=is_pdep(self.jobList))
-                supporting_info.append(job.supporting_info)
+                if hasattr(job,'supporting_info'):
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -274,33 +274,40 @@ def execute(self):
-                if len(row) > 1:  # monoatomic species have no frequencies nor rotational constants
-                    if isinstance(row[1].rotationalConstant.value, float):
+        if supporting_info:
+            #write supporting_info.csv for statmech jobs
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -269,13 +269,17 @@ def execute(self):
                 job.execute(output_directory=self.outputDirectory, plot=self.plot, pdep=is_pdep(self.jobList))
                 if hasattr(job,'supporting_info'):
                     supporting_info.append(job.supporting_info)
+                if hasattr(job,'raw_hindered_rotor_data'):
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -308,6 +312,13 @@ def execute(self):
                     atoms = ', '.join(["{0}    {1}".format(atom,"    ".join([str(c) for c in coords])) for atom, coords in zip(row[10], row[11])])
                     writer.writerow([label, row[1], row[2], row[3], rot, freq, row[7],row[8], row[9],atoms,row[12]])
+        if hindered_rotor_info:
+            hr_file = os.path.join(self.outputDirectory,'hindered_rotor_scan_data.csv')
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -308,6 +312,13 @@ def execute(self):
+            hr_file = os.path.join(self.outputDirectory,'hindered_rotor_scan_data.csv')
+            with open(hr_file, 'wb') as csvfile:
+                 writer = csv.writer(csvfile, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
+                 writer.writerow(['species','rotor_number','symmetry', 'spacing (radians)', ] + ['energy (cal/mol) {}'.format(i) for i in range(72)])
```
fixed

goldmanm commented on arkane/statmech.py:
```diff
@@ -457,26 +458,17 @@ def load(self, pdep=False, plot=False):
-                    if isinstance(scanLog, GaussianLog):
-                        scanLog.path = os.path.join(directory, scanLog.path)
+                    scanLog.path = os.path.join(directory, scanLog.path)
+                    if isinstance(scanLog, (GaussianLog,QChemLog)):
```
fixed

goldmanm commented on arkane/main.py:
```diff
@@ -308,6 +312,13 @@ def execute(self):
+            hr_file = os.path.join(self.outputDirectory,'hindered_rotor_scan_data.csv')
+            with open(hr_file, 'wb') as csvfile:
+                 writer = csv.writer(csvfile, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
+                 writer.writerow(['species','rotor_number','symmetry', 'spacing (radians)', ] + ['energy (cal/mol) {}'.format(i) for i in range(72)])
```
changed to resolution in degrees

goldmanm commented on arkane/main.py:
```diff
@@ -308,6 +312,13 @@ def execute(self):
+            hr_file = os.path.join(self.outputDirectory,'hindered_rotor_scan_data.csv')
+            with open(hr_file, 'wb') as csvfile:
+                 writer = csv.writer(csvfile, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
+                 writer.writerow(['species','rotor_number','symmetry', 'spacing (radians)', ] + ['energy (cal/mol) {}'.format(i) for i in range(72)])
```
I think this was arbitrary. I will rewrite it to first go through all hindered rotors, find the one with the smallest step size (and therefore longest length) and use that number.

goldmanm replied:
since most people run the same length for one paper, this should work out well.

goldmanm commented on arkane/statmech.py:
```diff
@@ -68,6 +68,9 @@ class ScanLog(object):
     scan energies.
+
+    Deprecated since the original Gaussian and Qchem scan logs are readable
+    with Arkane. Will be removed on or after RMG version 2.5
```
added

goldmanm commented on arkane/statmech.py:
```diff
@@ -198,6 +198,14 @@ def execute(self, output_directory=None, plot=False, pdep=False):
                 self.write_output(output_directory)
             except Exception as e:
                 logging.warning("Could not write statmech outputfile due to error: ".format(e))
+            if hasattr(self,'hindered_rotor_plots'):
```
fixed

goldmanm commented on arkane/statmech.py:
```diff
@@ -655,8 +666,19 @@ def plotHinderedRotor(self, angle, v_list, cosineRotor, fourierRotor, rotor, rot
+        ``rotor_[species_label]_0.pdf`` in the specified directory
+        """
+        if hasattr(self, 'hindered_rotor_plots'):
+            for fig, rotorIndex in self.hindered_rotor_plots:
```
fixed

goldmanm commented on arkane/statmech.py:
```diff
@@ -655,8 +666,19 @@ def plotHinderedRotor(self, angle, v_list, cosineRotor, fourierRotor, rotor, rot
+        if hasattr(self, 'hindered_rotor_plots'):
+            self.hindered_rotor_plots.append((fig,rotorIndex))
+        else:
+            self.hindered_rotor_plots = [(fig, rotorIndex)]
```
added and removed the hasattr statement

goldmanm commented on arkane/statmech.py:
```diff
@@ -198,6 +198,14 @@ def execute(self, output_directory=None, plot=False, pdep=False):
             except Exception as e:
                 logging.warning("Could not write statmech outputfile due to error: ".format(e))
+            if hasattr(self,'hindered_rotor_plots'):
+                try:
```
good idea. fixed

goldmanm commented on arkane/gaussian.py:
```diff
@@ -413,6 +413,67 @@ def loadScanEnergies(self):
+            while line != '':
+                if reached_input_spec_section:
+                    terms = line.split()
+                    if len(terms)==0:
```
added

goldmanm commented on arkane/gaussian.py:
```diff
@@ -413,6 +413,67 @@ def loadScanEnergies(self):
+                if reached_input_spec_section:
+                    terms = line.split()
+                    if len(terms)==0:
+                        #finished reading specs
```
added

goldmanm commented on arkane/gaussian.py:
```diff
@@ -413,6 +413,67 @@ def loadScanEnergies(self):
+                    elif terms[0] == 'B':
+                        action_index = 3 #bond length with 2 terms
+                    else:
+                        raise Exception('Unable to read scan specs for line: {}'.format(line))
```
Changed to ValueError and added more text 'This file has an option not supported by arkane.'

goldmanm commented:
I am a bit confused about what RMG-tests is telling me. It gives an error on `superminimal` of 

```
Regression input file not found. Not running a regression test.
The command "./run.sh rmg/superminimal no" exited with 0.
```

My intuition is that this is saying something errored when running superminimal, though when I try on my own computer, it works as expected.

mliu49 commented:
It's simply saying that it's not running the regression test (which compares model outputs) because there isn't an input file for it. This is separate from the RMG input file.

The failure is from the liquid oxidation example, which had a problem reading comments from the Chemkin file.

goldmanm commented:
I don't know why it's liquid oxidation given the error message has 'superminimal' in it's description.

mliu49 commented:
The error is on line 1538: https://travis-ci.org/ReactionMechanismGenerator/RMG-tests/jobs/543479094#L1538

goldmanm commented:
Thank you so much @mliu49! I'll figure it out now and get it fixed.

goldmanm commented:
I think I fixed the issue which dealt with parsing string comments. I took the file generated from the oxidation example, which contained the warning:

```
! This reaction has length 53 which is 1 longer than Chemkin allows.
! Consider shortening species names to run chemkin.
```

And tested it on chemkin, which gives no error. I even added 16 more characters, and still no error. I originally got the limit of 52 characters from an RMG post arguing we can only have 10 character long species because the reaction length cannot be longer than 52. I checked the recent version of chemkin documentation on Raineir 'input.pdf' and it gives no such restriction on character length.

Maybe in the last decade or so chemkin has become more flexible at parsing reactions, and it is no longer necessary to have this warning I wrote (or limit chemkin identifiers to 10 characters; we can use the full 16 characters). What do y'all think? I can post the extending chemkin identifiers to 16 characters as an issue and (for this PR) remove the warnings shown above?

mliu49 commented:
We currently have Chemkin 17.0, so the rmgpy.chemkin module is severely outdated...

That aside, it seems like the 16 character limit on species identifiers is still valid. However, reaction strings don't seem to have a limit as you mentioned.

I think it's fine to no longer check reaction line lengths. I'm still not sure how I feel about increasing species identifier lengths past 16 though...

mliu49 commented on rmgpy/chemkin.pyx:
```diff
@@ -73,10 +73,10 @@ def readThermoEntry(entry, Tmin=0, Tint=0, Tmax=0):
     object.
     
-    Format specification at http://www2.galcit.caltech.edu/EDL/public/formats/chemkin.html
+    Format specification at http://akrmys.com/public/chemkin/CKm_inp.html.en#THRM
```
I don't think this is a good link. It might be better to just cite the Chemkin Input Manual, even though it may not be accessible to people who don't have Chemkin.

goldmanm replied:
fixed

mliu49 commented on rmgpy/chemkin.pyx:
```diff
@@ -1567,13 +1588,15 @@ def writeReactionString(reaction, javaLibrary = False):
-    for an RMG-Java kinetics library.
+    for an RMG-Java kinetics library. `use_label` flag indicates that
+    the reaction string should use the species.label instead of the default RMG
+    species identifier.
```
I would consider the species label the default RMG species identifier. Suggestion:
```
Use `use_label=True` to force use of the default RMG species label instead 
of a ensuring a Chemkin compatible species identifier in the reaction string.
```

goldmanm replied:
fixed for all three instances

goldmanm commented:
I fixed the suggestions and added 2 more commits (one setting chemkin identifier length to 16 and one creating the `--no-plot` option)

mliu49 commented on rmgpy/chemkin.pyx:
```diff
@@ -73,10 +73,10 @@ def readThermoEntry(entry, Tmin=0, Tint=0, Tmax=0):
     """
     lines = entry.splitlines()
-    species = str(lines[0][0:18].split()[0].strip())
+    species = str(lines[0][0:22].split()[0].strip())
```
The thermo specification has columns 1-16 for the species name and columns 19-24 for a comment.

Was there a reason you chose 22? I think it could probably be safely changed to 24.

mliu49 replied:
Just to clarify a bit, if we're reading external chemkin files, we should still be able to parse the identifier and comment properly. If we're reading RMG chemkin files where we might have an extra long identifier, we might as well use the full 24 characters. If we're writing chemkin files, using 22 characters already breaks chemkin compatibility, so it won't matter if we use 24. So this can be changed at lines 1470 and 1968 as well.

goldmanm replied:
Good idea. I modified it to allow longer strings here.

mliu49 commented:
Just to double-check, I assume this is intended to be in the release?

goldmanm commented:
I would like for it to get incorporated in this release for my butanol paper, though I don't think it is worth it if all of us are not confident in it.

mliu49 commented:
I see. I think we can include it then, once we fix the RMG-tests build.

mliu49 commented:
I looked into why RMG-tests is failing. I think it's simply a bug related to how RMG-tests works.

The tests are failing on the regression test, which is run by an RMG-Py script. The script runs Cantera simulations using RMG-Py wrappers, which involves calling `getSpeciesIdentifer` at multiple points.

RMG-tests uses the testing version of RMG-Py to run this test (changed recently, but I think it would fail even if we were still using the benchmark version). Thus, when we try to simulate the benchmark model (with old identifiers) and call the new `getSpeciesIdentifier` function, we get a different identifier which does not match one in the model.

Therefore, I think we can actually ignore the RMG-tests failure for this PR. If everything else is ok, then I think we can merge this.

I only looked at the changes to chemkin.pyx, which I think is good now. @alongd, could you approve based on whether you think the Arkane changes are ok?

goldmanm commented:
I was thinking that was the issue with RMG-Tests but couldn't find exactly where the comparison was happening. I wonder if the cantera species & molecules are built from the rmg objects, instead of using the cti file, if this would prevent this error from appearing when changing `getSpeciesIdentifier`.

When I ran the regression test with RMG-Py (using the new version for both, which is redundant, i know), it worked perfectly fine.

alongd commented:
@goldmanm, I'm testing out this branch. Ran the Arkane species Benzyl example, and got:
```
(rmg_env) alongd@alongd:~/Code/RMG-Py/examples/arkane/species/Benzyl$ python $rmgpy/Arkane.py input.py
Using Theano backend.
Traceback (most recent call last):
  File "/home/alongd/Code/RMG-Py//Arkane.py", line 57, in <module>
    arkane.parseCommandLineArguments()
  File "/home/alongd/Code/RMG-Py/arkane/main.py", line 140, in parseCommandLineArguments
    self.plot = args.plot
AttributeError: 'Namespace' object has no attribute 'plot'
```
Were you able to run it smoothly?

goldmanm commented:
It was caused by changing from `--plot` to `--no-plot` since the method didnt specify a name. Added `dest='plot'`, and now functions as expected.

rwest commented:
Perhaps this discussion should be elsewhere, but we should decide the target of our CHEMKIN output. How many of our (desired) users are (for better or worse) using the models in hacked variants of Chemkin 2? 

Although maybe we should ignore such (hypothetical?) users and instead encourage them to use Cantera (which can have arbitrary long species names).

rwest commented:
P.s. I'm not sure the limitation on reaction string length on Chemkin was ever documented. Just figured out the hard way!

mliu49 commented:
@rwest, are you asking specifically about users who are using Chemkin II, or current versions of Chemkin/Chemkin-Pro?

I would think that no one still uses Chemkin II, and we should not worry about maintaining support for it at all.

I feel like there are still plenty of people who use Chemkin though (myself included). Chemkin files are also supported by other software (e.g. Chemical Workbench, Fluent).

I've also been told by others that the only way to get a Cantera input file is by writing it manually or converting from Chemkin, i.e. there isn't a way to dump Cantera species/reactions to the input file format. Idk if that will change with the new yaml format though.

alongd commented:
I'm running some examples, here are my initial impressions/comments (I haven't run reactions / network / explorer / ARC yet using this branch, but plan to):

- I see that the rotor scan PDF plot is not saved by default (I'm not specifying a plot flag when I run Arkane). I though we wanted to plot it by default.

- In the hindered_rotor_scan_data.csv file, are we sure the energy is in cal/mol as stated in the headers? I think the values we print there are actually J/mol (so either convert the values or change headers). Please double check.

- Regarding the geometry method in the SI file: Since we recommend passing Arkane the **freq** output file for the geometry when running rotors (for projection), the geometry method column actually reports the freq method. For example, if you run the benzyl and H examples in Arkane, both calculated at CBS-QB3, you get UB3LYP and CBS-QB3, respectively. I think Arkane might not have access to the geometry method anyway, so we might want to remove this column (BTW, just having the geometry method w/o the basis set isn't specific enough anyway)

- (not introduced by this PR, but related and let's fix it): We report unscaled frequencies. I think these are before projecting out the rotors, I think this should be mentioned in the SI file column header.

- The atom XYZ coordinates in the SI file is not in a multiline format (with line breaks) and not in a list format (has no `[` `]`). It has commas to separate the lines, and it cannot be directly pasted and used in an electronic structure software (or just be reported in the SI of a paper). Do you think we can save it as a multiline string? see [these](https://stackoverflow.com/questions/24018597/adding-a-newline-character-within-a-cell-csv) [two](https://stackoverflow.com/questions/566052/can-you-encode-cr-lf-in-into-csv-files) discussions.

rwest commented:
<img width="546" alt="Screen Shot 2019-06-13 at 1 22 46 PM" src="https://user-images.githubusercontent.com/93807/59453706-6ea3de80-8dde-11e9-8f86-99dce8078b04.png">

mliu49 commented:
Oh, interesting! I just learned that Chemkin II is essentially the open version of Chemkin. In that case, we should rethink this change...

goldmanm commented:
Thanks for looking and running the example files.

> I'm running some examples, here are my initial impressions/comments (I haven't run reactions / network / explorer / ARC yet using this branch, but plan to):
> 
>     * I see that the rotor scan PDF plot is not saved by default (I'm not specifying a plot flag when I run Arkane). I though we wanted to plot it by default.

When fixing up the code from some previous suggested changes, I accidentally removed the `if plot:` statement. Added it back and running the examples produces the plot.

>     * In the hindered_rotor_scan_data.csv file, are we sure the energy is in cal/mol as stated in the headers? I think the values we print there are actually J/mol (so either convert the values or change headers). Please double check.

Double checked and agree with you. Changed headers to J/mol.

>     * Regarding the geometry method in the SI file: Since we recommend passing Arkane the **freq** output file for the geometry when running rotors (for projection), the geometry method column actually reports the freq method. For example, if you run the benzyl and H examples in Arkane, both calculated at CBS-QB3, you get UB3LYP and CBS-QB3, respectively. I think Arkane might not have access to the geometry method anyway, so we might want to remove this column (BTW, just having the geometry method w/o the basis set isn't specific enough anyway)

I think I understand what you are saying, but correct me if I am wrong: We have an option in Arkane for the geometry file, but we tell users to put in the frequency file instead, so Arkane might not even have access to the geometry method at all, so it would be more accurate to call it 'frequency method' instead of 'geometry method'? And having the basis set would also be useful?

>     * (not introduced by this PR, but related and let's fix it): We report unscaled frequencies. I think these are before projecting out the rotors, I think this should be mentioned in the SI file column header.

Good point. I was thinking about calling them 'Raw Calculated Frequencies'. Thoughts about this wording?

>     * The atom XYZ coordinates in the SI file is not in a multiline format (with line breaks) and not in a list format (has no `[` `]`). It has commas to separate the lines, and it cannot be directly pasted and used in an electronic structure software (or just be reported in the SI of a paper). Do you think we can save it as a multiline string? see [these](https://stackoverflow.com/questions/24018597/adding-a-newline-character-within-a-cell-csv) [two](https://stackoverflow.com/questions/566052/can-you-encode-cr-lf-in-into-csv-files) discussions.

I think is possible to save it as a multiline string, though I am not sure that is necessarily better since it can introduce its own issues when reading and writing csv, which I why I converted it into this format. According to [Wikipedia "many CSV implementations do not support embedded line breaks"](https://en.wikipedia.org/wiki/Comma-separated_values). 

I also was not trying to put it directly in a format which can be directly used in electronic structure software because there is no standard format for all of electronic structure software, so the user may have to convert it anyways.

It is not super complicated to convert between formats given the current structure For example, the user can easily convert this to, per se, the Gaussian input format by tying `s.replace(', ','\n')` and then printing the result, which can be used in Gaussian or the SI of a paper.

I think finding work around to have line breaks will likely introduce more errors than it is solves, especially given the lack of universal support for line breaks in CSV files.

alongd commented:
I agree - introducing line breaks will just cause more issues... :) Let's keep the XYZ as is.

I recommend avoiding reporting the method (and basis set). It's not always the opt level, and it's not always the freq level either.

For the frequencies header, what do you say about `Calculated Frequencies (unscaled and prior to projection, cm^-1)`?

goldmanm commented:
@mliu49, not allowing our output to be used by a chemkin version which is still widely used would not be idea. Some ideas:

1. go back to species of length 10
2. give an option to use length 10 species in the rmg input file, defaults to 16. This isn't that useful if someone takes a published mechanism from our group and tries to put it in chemkin II. 
3. keep length at 16. Since chemkin II is open and editable, inform users that RMG is fully complient with chemkin II input specifications, though some versions of chemkin II are limited by reaction length, and that they may need to modify chemkin II reaction parsing to use RMG's output.

mliu49 commented:
I think we should go undo the changes and go back to current behavior. We can/should have a more in-depth discussion about it later.

Do you still want to keep the override to print the full species labels? I think I would prefer to move that out of this PR for more discussion.

mliu49 commented:
Final decision is that this PR will not be in v2.4.0.

alongd commented:
Some impressions from running our examples:

- Species:
  - The supporting info file still has the geometry method. Didn't we want to remove it? Or was the code implementation changed?
  - In the SI file we have `Electronic energy (J/mol)`, `Energy with atom corrections (J/mol)`, and  `Energy including zero point (J/mol)`. I think the titles aren't entirely clear. Does `Energy with atom corrections (J/mol)` include BAC? Does `Energy including zero point (J/mol)` include the former corrections as well? I think the answers to both Qs is (or should be) yes, but the titles don't give that away. Thinking of what users might want (correct me please) I would suggest to give E_elect, E0, and E0 with corrections, with the following corresponding titles: `Electronic energy (J/mol)`, `E0 (electronic energy + ZPE, J/mol)`, and  `E0 with atom and bond corrections (J/mol)`. What do you think?
  - How can we use the `hindered_rotor_scan_data.csv` file in a species input file?
  - Would we like to convert all J/mol to kJ/mol to make the values more readable? We could also restrict printing kJ/mol to only 3 decimal figs after the point (relates to rotors and SI files)
- Reactions:
  - I compared the output for the 23dimethylpropoxy example on this branch and on master, there were some differences in rate (not too significant). I think they'll disappear once this branch is rebased, so just a note to ourselves to check again after rebasing to make sure we didn't introduce it here.
  - The supporting info file when running a reaction is very different than for species (seems like it's the old version, e.g., I'm getting `Unscaled frequencies (cm-1)` instead of `Calculated Frequencies (unscaled and prior to projection, cm^-1)` and don't see the new attributes). Seems like the SI file for reactions is processed differently and wasn't updated. (I observed the same when running networks and explorer, which is kind of a network)

I plan to test this branch with ARC, it's likely though that I'll implement the changes, if any, in ARC, but there might be minor comments to implement here as well.

goldmanm commented:
Species:
        The supporting info file still has the geometry method. Didn't we want to remove it? Or was the code implementation changed?

> Yes. I have removed it from the supporting info file.

        In the SI file we have Electronic energy (J/mol), Energy with atom corrections (J/mol), and Energy including zero point (J/mol). I think the titles aren't entirely clear. Does Energy with atom corrections (J/mol) include BAC? Does Energy including zero point (J/mol) include the former corrections as well? I think the answers to both Qs is (or should be) yes, but the titles don't give that away. Thinking of what users might want (correct me please) I would suggest to give E_elect, E0, and E0 with corrections, with the following corresponding titles: Electronic energy (J/mol), E0 (electronic energy + ZPE, J/mol), and E0 with atom and bond corrections (J/mol). What do you think?

> Good point. I used the three values that Arkane natively uses, which might not be the most useful for the user. The answer to these questions is yes. I changed the definitions and names to reflect your suggestions.

        How can we use the hindered_rotor_scan_data.csv file in a species input file?

> Underlying this question is an assumption that it is necessary and or desirable to use the output from an Arkane species run as an input into an Arkane species run. To me, this seems unnecessary. Since by virtue of already having run Arkane, there must have been a hindered rotor log file. This same file can be used the next time Arkane is run. Therefore, there isn't a need to run Arkane using the output from hindered_rotor_scan_data.csv. Maybe I am missing some other use case where the hindered rotor log file disappears but the user manages to have hindered_rotor_scan_data.csv. It seems like this can happen if someone wants to replicate another person's publication, which in this scenario, I think it is reasonable to expect them to copy the values into a text file and read it in with ScanLog (or better yet, write a script to do that). I imagine this is what they'd have to do before this pull request, anyways. Let me know if I am missing something.

        Would we like to convert all J/mol to kJ/mol to make the values more readable? We could also restrict printing kJ/mol to only 3 decimal figs after the point (relates to rotors and SI files)

> If readability is desired, the user can view values in the `output.py` file. When I asked users of arkane what they used the `output.py` file for, they mentioned to look up information. That is why that file provides tables of reaction rates, heat capacities, entropies, etc., at various temperatures, and other information useful for humans. The output.py file already outputs energies in kJ/mol with three decimal places. Much of this information in the `output.py` format is not machine readable, making it a hassle to present it as supplemental information for a manuscript or use it in another quantum job. The intention of the csv file is to provide information in a machine readable format, where the users do not need to have the correct version RMG fully installed.

    Reactions:

        I compared the output for the 23dimethylpropoxy example on this branch and on master, there were some differences in rate (not too significant). I think they'll disappear once this branch is rebased, so just a note to ourselves to check again after rebasing to make sure we didn't introduce it here.

> Rebasing seems to have fixed this issue: I ran on both branches and compared `output.py`, `chem.inp`, and `arkane.log`. The only difference was in the log file, which differed in the time of initiation and termination and the memory used (which are expected).

        The supporting info file when running a reaction is very different than for species (seems like it's the old version, e.g., I'm getting Unscaled frequencies (cm-1) instead of Calculated Frequencies (unscaled and prior to projection, cm^-1) and don't see the new attributes). Seems like the SI file for reactions is processed differently and wasn't updated. (I observed the same when running networks and explorer, which is kind of a network)

> When I ran the 23dimethylpropoxy example on this branch after rebasing, I got a full table for supporting_info.csv. (I didn't try it before rebasing) The only thing missing was the T1 diagnostic, which is expected because Qchem did not output that diagnostic. Double check with the rebased version and let me know if you're still experiencing it.

goldmanm commented:
Some changes to this branch:

1. moved species length changes to #1648 
2. removed the need to specify geometry file if it is identical to the frequency file geometry
2. removed the need to specify bonds if the structure of the molecule is given
2. refactored applyEnergyCorrections into two methods, one for bonds and one for atoms
2. added the D1 diagnostic to the supporting_information.csv output by RMG

I won't add anything else to this PR, I promise.

alongd commented on documentation/source/users/arkane/input.rst:
```diff
@@ -204,7 +204,8 @@ Allowed bond types for the ``bonds`` parameter are, e.g., ``'C-H'``, ``'C-C'``,
 ``'O=S=O'`` is also allowed.
 
-The order of elements in the bond correction label is not important. Use ``-``/``=``/``#`` to denote a
+The order of elements in the bond correction label is important. The first atom
```
[This line](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/master/arkane/statmech.py#L943) also checks the reverse of the string for the bond correction. I thought it worked fine. Did you experience problems reading bond corrections in reverse? If so, let's fix it (but keep the documentation)

goldmanm replied:
If this is the case then, I should change [this line](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/master/arkane/statmech.py#L943) to avoid confusion. 

It is minor, but one bond correction for 'cbs-qb3-paraskevas' leads to unexpected behavior where reversing the order doesn't match:

```
>>> symbol = 'Cl-C'
>>> symbol[::-1]
'C-lC'
```

Should I change the commented line with the incorrect information?

alongd replied:
Yes, this is indeed a bug. Could you change it so it works for atoms with more than one char long symbols? Good catch

goldmanm replied:
It seems like @cgrambow's PR already fixed this issue (using algorithm from the updated master branch):

```
>>> import re
>>> symbol = 'C=Cl'
>>> ''.join(re.findall('[a-zA-Z]+|[^a-zA-Z]+', symbol)[::-1])
'Cl=C'
```

goldmanm commented on arkane/input.py:
```diff
@@ -664,8 +664,8 @@ def loadInputFile(path):
             job.modelChemistry = sp_level
             job.frequencyScaleFactor = frequency_scale_factor
             job.includeHinderedRotors = use_hindered_rotors
-            job.applyAtomEnergyCorrections = use_atom_corrections
```
I am fine dropping this. @cgrambow's changes are much better.

goldmanm commented on arkane/statmech.py:
```diff
@@ -347,6 +339,15 @@ def load(self, pdep=False, plot=False):
             statmechLog.path = os.path.join(directory, statmechLog.path)
+        try:
+            geomLog = local_context['geometry']
+            if isinstance(geomLog, Log) and not isinstance(energy, (GaussianLog, QChemLog, MolproLog)):
```
Haha. Not from this PR, but definitely written by me. This checks that the log file is of the `Log` class (which doesn't contain usable methods) but not of any of its descended classes (which has specific methods). If this is the case, then we have to determine which software generated the log file and create that new object ( which is what the following line does).

Do you know if there's a better way to check that something is an instance of a class but not of any of its descendant classes? I didn't find a way when writing this code.

alongd replied:
Gotcha. Maybe `isinstane(geomLog, Log) and not issubclass(geomLog, Log)` (I haven't tested)? We'll soon incorporate (at least plan to) support for TeraChem and Orca, let's minimize the places we need to modify when adding new software support.

goldmanm replied:
Excellent suggestion. I implemented this in all 6 places it occurs in statmech.py, and made a separate commit, because it impacts sections not previously touched with this PR.

goldmanm replied:
I changed this from the original implementation because `issubclass` requires both attributes to be classes (not instances) and since a class will return itself as true:

```
import numpy as np

issubclass(np.ndarray, np.ndarray)
Out[2]: True
```

The new implementation uses `isinstance (geomLog, Log) and type(geomLog).__name__ != 'Log'`. This should do the same thing we originally wanted to use `issubclass` for.

goldmanm commented on arkane/gaussian.py:
```diff
@@ -468,3 +468,30 @@ def loadNegativeFrequency(self):
             raise Exception('Unable to find imaginary frequency of {1} '
                             'in Gaussian output file {0}'.format(self.path, self.species.label))
         return frequency
+
```
We are not using this because, as you mentioned, we removed the commit to print out level of theory. I will remove this commit too. I initially left it because I am thinking of fixing it, adding a method to find basis set, and do the same for energy calculations, if/when I have time.

goldmanm commented on arkane/statmech.py:
```diff
@@ -87,6 +91,8 @@ class ScanLog(object):
     def __init__(self, path):
         self.path = path
+        warnings.warn("The option to read a scan log from a tabulated text file"
+                      "is no longer supported and may be removed after Version: 2.5 ", DeprecationWarning)
```
I can remove the deprecation warning. Could you write the documentation in a separate PR 
describing to users how to create and modify these scan files?

alongd replied:
Turns out it is already documented. I elaborated it a bit, see #1650

goldmanm commented on arkane/statmech.py:
```diff
@@ -558,6 +551,37 @@ def load(self, pdep=False):
+            self.supporting_info.append(None)
+        self.supporting_info.append(e_electronic)
+        self.supporting_info.append(e_electronic + zpe)
+        self.supporting_info.append(e0)
```
based on the relabeling recently done, I believe e0 = e_electronic + corrections + zpe, which didn't meet the criteria you wanted e to output (having one without corrections).

goldmanm commented on arkane/main.py:
```diff
@@ -125,7 +125,7 @@ def parseCommandLineArguments(self):
 
         # Add options for controlling generation of plots
-        parser.add_argument('-p', '--plot', action='store_true', default=True, help='generate plots of results')
+        parser.add_argument('-p', '--plot', action='store_true', default=False, help='generate plots of results')
```
I kept them separate because they fix two different issues. I will squash them.

goldmanm commented on arkane/thermo.py:
```diff
@@ -181,26 +195,31 @@ def save(self, outputFile):
+                    try:
+                        elementCounts = species.props['elementCounts']
+                    except KeyError:
+                        elementCounts = {'C': 0, 'H': 0}
```
We definitely could, and maybe should. This PR didn't change this functionality, so I'll leave it to another PR to make this change.

alongd replied:
No problem

alongd replied:
Done in #1651

goldmanm commented on rmgpy/pdep/configuration.pyx:
```diff
@@ -275,10 +282,18 @@ cdef class Configuration:
+                    modes.insert(0, IdealGasTranslation(mass=(mu/constants.amu,"amu")))
+                else:
+                    if len(mass) != 3:
+                        raise AttributeError('Length of masses should be three for termolecular reactants.')
```
Can and will do.

goldmanm commented on rmgpy/tools/merge_models.py:
```diff
@@ -122,10 +122,21 @@ def execute(inputModelFiles, **kwargs):
         finalModel = finalModel.merge(model)
         Nspec = len(finalModel.species)
         Nrxn = len(finalModel.reactions)
-        print 'Added {1:d} out of {2:d} ({3:.1f}%) unique species from model #{0:d}.'.format(i+1, Nspec - Nspec0, len(model.species), (Nspec - Nspec0) * 100. / len(model.species))
```
Nope (especially after modifying other parts of merge_models) and I have removed it.

goldmanm commented on arkane/main.py:
```diff
@@ -301,10 +301,9 @@ def execute(self):
-                # if there is an explorer job the pdep job will be run in the explorer job
+                job.execute(output_directory=self.outputDirectory, plot=self.plot)
+            elif isinstance(job, PressureDependenceJob) and not any([isinstance(job, ExplorerJob) for job in self.jobList]):
+                #if there is an explorer job the pdep job will be run in the explorer job
```
Me too. Thanks for catching this. It seems like it must have slipped through while dealing with merge conflicts for all the PEP-8 changes from that PR. Fixed.

goldmanm commented:
@alongd, is there anything you'd still like changed about this PR?

alongd commented:
@goldmanm, Thanks for this important and extensive contribution. Sorry for being slow in reviewing.
I have just a minor comment: could you please update the docstrings of functions / methods where the arguments changed? For example, in `KineticsJob.execute()` the first argument is now `output_directory` rather than `outputFile` (other than name change, the content has changed as well).
After that, please rebase (and squash the docstring changes), I think we're ready to merge (I'll update ARC in parallel to merging this PR)

goldmanm commented:
I improved the docstrings for the methods in the thermo, kinetics, and statmech files for their `execute`, `save_output` and `save_chemkin` methods. Examples below:

```
        Execute the thermodynamics job, saving the results within
        the `output_directory`.

        If `plot` is true, then plots of the raw and fitted values for heat
        capacity, entropy, enthalpy, gibbs free energy, and hindered rotors
        will be saved.
```

```
        Save the results of the thermodynamics job to the `output.py` file located
        in `output_directory`.
```

```
        Appends the thermo block to `chem.inp` and species name to
        `species_dictionary.txt` within the `outut_directory` specified
```
```
